### PR TITLE
Added asynchronous transfer support to the SPI display interface

### PR DESCRIPTION
**Pull request by mikeoliphant**
## Description
This PR adds asynchronous transfer support to the SPI display interface. This increases the throughput of SPI-based display updates by almost 50%.

## Motivation and Context
Previously, SPI display data transfers were synchronous. The meant that any other work done by the display transfer code happened in between transfers. The display code does a copy operation and byte-switches the 16bit data as required by the SPI displays. This copy/swap operation takes a significant amount of time. With asynchronous operation, the copy/swap can be done while the previous transfer is occurring.

Two methods have been added to DisplayInterface: SendData16Windowed and FillData16. These functions remove significant duplicate code among the SPI display drivers and put in one place. They also allow for the asynchronous transfer to be implemented.

The display drivers updated are: **ILI9341_240x320_SPI**, **ILI9342_320x240_SPI**, **SSD1331_94x64_SPI**, **ST7735S_SPI**, and **ST7789V_240x320_SPI**.

There are two other display drivers that were not modified. **SSD1306_128x64** is a 1-bit display which requires custom data handling and **Otm8009a_DSI_Video_Mode** uses a different display transfer implementation (**DSI_To_Display_Video_Mode.cpp**).

## How Has This Been Tested?
Tested locally on M5Stack Core2.

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dev Containers (changes related with Dev Containers, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [ ] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).

### CODE REVIEW

File: src/nanoFramework.Graphics/Graphics/Displays/Spi_To_Display.cpp [hunk 1]
```diff
@@ -158,24 +167,137 @@ void DisplayInterface::DisplayBacklight(bool on) // true = on
...
+{
+    while (length > 0)
+    {
+        CLR_UINT32 toWrite = length;
+
+        if (toWrite > (SPI_MAX_TRANSFER_16 - bufferWritten))
+            toWrite = (SPI_MAX_TRANSFER_16 - bufferWritten);
+
+        // If we need to byteswap the data, do that - otherwise we can just memcpy it
+        if (doByteSwap)
```
Comment (josesimoes) [inline src/nanoFramework.Graphics/Graphics/Displays/Spi_To_Display.cpp:234]: open and close brackets please

File: src/nanoFramework.Graphics/Graphics/Displays/Spi_To_Display.cpp [hunk 2]
```diff
@@ -158,24 +167,137 @@ void DisplayInterface::DisplayBacklight(bool on) // true = on
...
+    }
+}
+
+void SendData16(CLR_UINT16 *data, CLR_UINT32 length, bool doByteSwap)
+{
+    while (length > 0)
+    {
+        CLR_UINT32 toWrite = length;
+
+        if (toWrite > (SPI_MAX_TRANSFER_16 - bufferWritten))
```
Comment (josesimoes) [inline src/nanoFramework.Graphics/Graphics/Displays/Spi_To_Display.cpp:230]: here too

**Review by josesimoes — COMMENTED** [review]
LGTM! Just a couple of nitpicks 😅

**Review by josesimoes — APPROVED** [review]
LGTM!

**Comment by mikeoliphant** [comment]
I'm mostly using my own code for testing. I've been thinking about putting together some more formal low-level graphics test cases. Not sure where they should live though - maybe in nanoFramework.Graphics?

**Comment by josesimoes** [comment]
Yes that looks like a good location for those.
Please add a new folder in the root and organize them as you see fit.

Regarding the M5stack , maybe you want to add a few lines to the existing test apps that are already there.

**Comment by mikeoliphant** [comment]
Yes - I was thinking to have some low-level tests in nanoFramework.Graphics that could then be referenced by the individual device test apps.

**Pull request merged**